File: src/684_redundant_connection/redundant_connection.cpp
#include <iostream>
#include <vector>
using namespace std;

class DSU {
   public:
    vector<int> parent, rank;
    DSU(int n) : parent(n), rank(n, 0) {
        for (int i = 0; i < n; ++i) parent[i] = i;
    }
    int find(int node) {
        if (parent[node] != node) parent[node] = find(parent[node]);
        return parent[node];
    }
    bool union_(int node1, int node2) {
        int parent1 = find(node1), parent2 = find(node2);
        if (parent1 == parent2) return false;
        if (rank[parent1] < rank[parent2])
            parent[parent1] = parent2;
        else if (rank[parent2] < rank[parent1])
            parent[parent2] = parent1;
        else {
            parent[parent1] = parent2;
            rank[parent1]++;
        }
        return true;
    }
};

class Solution {
   public:
    vector<int> findRedundantConnection(vector<vector<int>>& edges) {
        int n = edges.size();
        DSU dsu(n);
        for (const auto& edge : edges) {
            int src = edge[0] - 1, dst = edge[1] - 1;
            if (!dsu.union_(src, dst)) return {edge[0], edge[1]};
        }
        return {};
    }
};

// Test scaffolding
void printVec(const vector<int>& v) {
    cout << "[";
    for (size_t i = 0; i < v.size(); ++i) {
        cout << v[i];
        if (i + 1 < v.size()) cout << ", ";
    }
    cout << "]";
}

void test() {
    Solution sol;
    vector<vector<int>> e1 = {{1, 2}, {1, 3}, {2, 3}};
    printVec(sol.findRedundantConnection(e1));
    cout << " | [2,3]\n";
    vector<vector<int>> e2 = {{1, 2}, {2, 3}, {3, 4}, {1, 4}, {1, 5}};
    printVec(sol.findRedundantConnection(e2));
    cout << " | [1,4]\n";
    vector<vector<int>> e3 = {{1, 2}, {2, 3}, {3, 1}};
    printVec(sol.findRedundantConnection(e3));
    cout << " | [3,1]\n";
    vector<vector<int>> e4 = {{1, 2}, {2, 3}, {3, 4}, {4, 1}, {1, 5}};
    printVec(sol.findRedundantConnection(e4));
    cout << " | [4,1]\n";
}

int main() {
    test();
    return 0;
}

File: src/684_redundant_connection/redundant_connection.py
from typing import List


class DSU:
    def __init__(self, n: int):
        self.parent = list(range(n))
        self.rank = [0] * n

    def find(self, node: int) -> int:
        if self.parent[node] != node:
            self.parent[node] = self.find(self.parent[node])

        return self.parent[node]

    def union(self, node1, node2):
        parent1 = self.find(node1)
        parent2 = self.find(node2)

        if parent1 == parent2:
            return False

        if self.rank[parent1] < self.rank[parent2]:
            self.parent[parent1] = parent2
        elif self.rank[parent2] < self.rank[parent1]:
            self.parent[parent2] = parent1
        else:
            self.parent[parent1] = parent2
            self.rank[parent1] += 1

        return True


class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)

        dsu = DSU(n)

        for src, dst in edges:
            if not dsu.union(src - 1, dst - 1):
                return [src, dst]

        return []


# Test scaffolding
def test():
    sol = Solution()
    print(sol.findRedundantConnection([[1, 2], [1, 3], [2, 3]]), "| [2,3]")
    print(
        sol.findRedundantConnection([[1, 2], [2, 3], [3, 4], [1, 4], [1, 5]]), "| [1,4]"
    )
    print(sol.findRedundantConnection([[1, 2], [2, 3], [3, 1]]), "| [3,1]")
    print(
        sol.findRedundantConnection([[1, 2], [2, 3], [3, 4], [4, 1], [1, 5]]), "| [4,1]"
    )


if __name__ == "__main__":
    test()
